### DISCUSSION TITLE
How to inject a NestJS provider in MikroORM's onDelete function?

### DISCUSSION THREAD

**Discussion by Alwinator** [mikro-orm/mikro-orm#3314] [Q&A] [answered]
What is the best way to fix this problem / error:

```ts
@Entity()
export class MyEntity {
  constructor(
    private readonly bucket: MyS3Bucket
  ) {
    super()
    console.log(bucket) // bucket is NOT undefined
    this.deleteFromS3.bind(this)
  }

  // ...

  @BeforeDelete()
  async deleteFromS3() {
    console.log(this.bucket) // bucket is undefined
    await this.bucket.deleteObjectsOfEntity(this) // error because bucket is undefined
  }
}
```

**Comment by B4nan** [comment, marked as answer]
The ORM never uses entity constructors for managed entities, its only used for new ones (those that will issue insert query) - I doubt you are trying to remove those (as they dont even exist in database).

**Reply by Alwinator** [reply]
Do you have any idea how I could solve my problem then? I have both a Postgres Database and S3 object storage. When I upload an image the file is uploaded to S3 and the metadata is inserted in the Postgres DB using MikroORM. My goal is to automatically delete the file from S3 when the metadata is deleted in the MikroORM database.

**Reply by B4nan** [reply]
You need to get that dependency from somewhere. You could pass it to the entity via EM filter params, setting it somewhere via `em.getFilterParams()`, and in the hook you can set first `args` parameter and get that via `args.em.getFilterParam()`.

**Reply by B4nan** [reply]
Note that filter params are cloned on forking, so it might not be suitable for some internal resources like database handles. But you could monkey patch the `args.em` or do something like that to get around it, no need to use the filter params, its just a type safe way.

**Reply by Alwinator** [reply]
I implemented the way with filter queries. (See code below). The bucket variable is not undefined and is a MyS3Bucket instance. However, I am getting the following error:
```
TypeError: The "key" argument must be of type string or an instance of ArrayBuffer, Buffer, TypedArray, DataView, KeyObject, or CryptoKey. Received undefined
    at new NodeError (node:internal/errors:371:5)
    at prepareSecretKey (node:internal/crypto/keys:570:11)
    at new Hmac (node:internal/crypto/hash:132:9)
    at Object.createHmac (node:crypto:162:10)
    at Object.hmac (/home/alwin/WebStormsProjects/myproject/node_modules/aws-sdk/lib/util.js:436:30)
    at Object.getSigningKey (/home/alwin/WebStormsProjects/myproject/node_modules/aws-sdk/lib/signers/v4_credentials.js:62:8)
    at V4.signature (/home/alwin/WebStormsProjects/myproject/node_modules/aws-sdk/lib/signers/v4.js:98:36)
    at V4.authorization (/home/alwin/WebStormsProjects/myproject/node_modules/aws-sdk/lib/signers/v4.js:93:36)
    at V4.addAuthorization (/home/alwin/WebStormsProjects/myproject/node_modules/aws-sdk/lib/signers/v4.js:35:12)
    at /home/alwin/WebStormsProjects/myproject/node_modules/aws-sdk/lib/event_listeners.js:282:18
```

I think the error is caused because some variables get lost while cloning (as you mentioned)

```ts
export class MyService{
  constructor(
    private readonly bucket: MyS3Bucket,
    private readonly em: EntityManager
  ){
    em.setFilterParams('bucket', bucket)
  }
```
```ts
@Entity()
export class MyEntity {
  @BeforeDelete()
  async deleteFromS3(args: any) {
    const bucket: MyS3Bucket = args.em.getFilterParams('bucket')
    await this.bucket.deleteObjectsOfEntity(this)
  }
}
```
You also mentioned that there is a second solution with monkey patching. I am not sure how I should do that. Could you please provide a code snippet or a more detailed description?

**Reply by B4nan** [reply]
You can do things like `(em.getDriver() as any).foo = 123` and read that later. I used the driver here on purpose, as thats singleton, so you get the same value even for differnet EM forks.

**Reply by Alwinator** [reply]
That worked perfectly, thank you very much! ;)
